File: c/tests/03_setters.cpp
// SPDX-License-Identifier: MIT
// Copyright (c) 2021-2022 brainpower <brainpower at mailbox dot org>

#include "test.hpp"

TEST_CASE("setters: positional argument help", "setters") {
  vector<const char *> argv = {
    "/test02",
  };

  CheckArgUPtr ca(
    checkarg_new(argv.size(), (char **)argv.data(), "test02", NULL, NULL),
    &checkarg_free);

  checkarg_add_autohelp(ca.get());
  checkarg_add(ca.get(), 'i', "input", "file to read from", CA_VT_REQUIRED, NULL);
  checkarg_add(ca.get(), 't', "test", "run tests", CA_VT_REQUIRED, NULL);

  checkarg_set_posarg_help(ca.get(), "[files...]", "one or more output files");

  int ret = checkarg_parse(ca.get());

  CHECK(ret == CA_ALLOK);
  string usage, help;
  {
    unique_ptr<char> xusage(checkarg_usage(ca.get()));
    unique_ptr<char> xhelp(checkarg_autohelp(ca.get()));
    usage = xusage.get();
    help  = xhelp.get();
  }
  REQUIRE(usage == "Usage: test02 [options] [files...]");
  REQUIRE(help ==
    "Usage: test02 [options] [files...]\n"
    "\n"
    "Options:\n"
    "   -h, --help   show this help message and exit\n"
    "   -i, --input  file to read from\n"
    "   -t, --test   run tests\n"
    "\n"
    "Positional Arguments:\n"
    "one or more output files\n"
  );
}

TEST_CASE("setters: override usage line", "setters") {
  vector<const char *> argv = {
    "/test02",
  };

  CheckArgUPtr ca(
    checkarg_new(argv.size(), (char **)argv.data(), "test02", NULL, NULL),
    &checkarg_free);

  checkarg_add_autohelp(ca.get());
  checkarg_add(ca.get(), 'i', "input", "file to read from", CA_VT_REQUIRED, NULL);
  checkarg_add(ca.get(), 't', "test", "run tests", CA_VT_REQUIRED, NULL);

  checkarg_set_usage_line(ca.get(), "this is a really stupid usage line");

  int ret = checkarg_parse(ca.get());

  CHECK(ret == CA_ALLOK);

  string usage, help;
  {
    unique_ptr<char> xusage(checkarg_usage(ca.get()));
    unique_ptr<char> xhelp(checkarg_autohelp(ca.get()));
    usage = xusage.get();
    help  = xhelp.get();
  }
  REQUIRE(usage == "Usage: this is a really stupid usage line");
  REQUIRE(help ==
    "Usage: this is a really stupid usage line\n"
    "\n"
    "Options:\n"
    "   -h, --help   show this help message and exit\n"
    "   -i, --input  file to read from\n"
    "   -t, --test   run tests\n"
  );
}

File: rust/tests/03_setters.rs
// SPDX-License-Identifier: MIT
// Copyright (c) 2021-2022 brainpower <brainpower at mailbox dot org>
#![feature(assert_matches)]

#[cfg(test)]
mod tests {
  use checkarg::{CheckArg, ValueType, RC};
  use std::assert_matches::assert_matches;

  #[test]
  fn positional_argument_help() {
    let argv = vec!["/test02"];

    let mut ca = CheckArg::new("test02");

    ca.add_autohelp();
    ca.add(
      'i',
      "input",
      "file to read from",
      ValueType::Required,
      Some(""),
    );
    ca.add('t', "test", "run tests", ValueType::Required, Some(""));

    ca.set_posarg_help("[files...]", "one or more output files");

    let rc = ca.parse(&argv);

    assert_matches!(rc, RC::Ok, "parsing failed");

    assert_eq!(ca.usage(), "Usage: test02 [options] [files...]");
    assert_eq!(
      ca.autohelp(),
      "\
Usage: test02 [options] [files...]

Options:
   -h, --help   show this help message and exit
   -i, --input  file to read from
   -t, --test   run tests

Positional Arguments:
one or more output files\n"
    );
  }

  #[test]
  fn override_usage_line() {
    let argv = vec!["/test02"];

    let mut ca = CheckArg::new("test02");

    ca.add_autohelp();
    ca.add(
      'i',
      "input",
      "file to read from",
      ValueType::Required,
      Some(""),
    );
    ca.add('t', "test", "run tests", ValueType::Required, Some(""));

    ca.set_usage_line("this is a really stupid usage line");

    let rc = ca.parse(&argv);

    assert_matches!(rc, RC::Ok, "parsing failed");

    assert_eq!(ca.usage(), "Usage: this is a really stupid usage line");
    assert_eq!(
      ca.autohelp(),
      "\
Usage: this is a really stupid usage line

Options:
   -h, --help   show this help message and exit
   -i, --input  file to read from
   -t, --test   run tests\n"
    );
  }
}
